### PR TITLE
UDP support

### PR DESCRIPTION
**Pull request by mpl**
### What does this PR do?

This PR adds limited support for UDP networking in Traefik.
Limited because:
- No TLS for now, which means
- No routing rule, since there is no criterion on which to match packets (there is no PathPrefix notion since there are no requests at the transport layer level, and no HostSNI since no TLS). Which means it's pretty much only load-balancing in one form or another at the moment.
- file provider only for now.

Remark: we adopted the docker syntax to distinguish TCP (HTTP) VS UDP for the entrypoints configuration:

```toml
[entryPoints]
  [entryPoints.udp]
    address = ":8093/udp"
  [entryPoints.http]
    address = ":8093" # same as ":8093/tcp"
```

### Motivation

Addresses #5048 

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### Additional Notes

Co-authored-by: Julien Salleyron <julien.salleyron@gmail.com>

### CODE REVIEW

**Comment by nxadm** [comment]
I don't think my use case would work in this setup, but it may be useful when looking for real time uses.

We use Netscaler for UDP load balancing for RADIUS containers (Network authentication). While simple RADIUS authentications don't need backend persistence (= related requests must be handled by the same backend), everything that uses encryption does (e.g. PEAP and EAP-TLS): it's a long exchange of certificate information. The persistence is done by looking at the (unencrypted) "User-Name" attribute in the RADIUS UDP packages and sending them to the same backend.

**Comment by mpl** [comment]
> I don't think my use case would work in this setup, but it may be useful when looking for real time uses.
> 
> We use Netscaler for UDP load balancing for RADIUS containers (Network authentication). While simple RADIUS authentications don't need backend persistence (= related requests must be handled by the same backend), everything that uses encryption does (e.g. PEAP and EAP-TLS): it's a long exchange of certificate information. The persistence is done by looking at the (unencrypted) "User-Name" attribute in the RADIUS UDP packages and sending them to the same backend.

Hi @nxadm ,

We already handle some form of what we call a session persistence (based on IP:port), simply because a proxy must be able to "remember" where (i.e. to which client) it should send a response from a backend server. As you might have guessed, this session expires in case of inactivity (timeout) after a duration, which is hardcoded for now (to 3 seconds, we'll add a note about it soon in another commit on this PR). If we understand your problem correctly, we think it might just be enough to tweak this timeout to fit your needs, if it turns out that your session exchanges have idle periods of more than 3 seconds. We most likely intend to someday make this timeout configurable, but we were waiting to get more usage feedback (such as yours) to think further about it

**Comment by nxadm** [comment]
@mpl Logins take normally a lot less than 3 seconds. However, in the RADIUS world besides external resources (typically LDAP or DBs), it's very common to proxy request to their authoritative RADIUS server, so in some cases the chain may be long* and the timeout exceeded.

But in the case in this protocol, and more importantly I suspect there are others, the issue is not timeout for a working load-balancing because IP:port don’t suffices. So in an ideal world, the criterium on which the loadbalancing is done should be pluggable, so poor souls like me would try to implement the support for their specific protocols :).

In the case of RADIUS this is the use case:
“Classic" clients (user devices) don't initiate a connection with the RADIUS server: there is always an intermediate that network-wise are the source: switches (cabled connections), Wireless Access Controllers (WiFi) and other RADIUS servers (remote users, decision points). Session persistence (based on IP:port) would in this case result in a load balancing scenario where users are not spread evenly across the backends. Some backends will only serve a few dozens connections a day, e.g. from a wired switch. Other backends would get hundreds connections per seconds, e.g. from an other RADIUS server or a Wireless Access Controller. The only way to load balance in this case is looking at the User-Name attribute in the RADIUS UDP packet. 

*: thin of [eduroam](https://www.eduroam.org/) where researchers/students can connect to a wifi abroad and the authentication is done on their institution.

File: pkg/server/router/udp/router.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		entryPointHandlers[entryPointName] = handler
+	}
+	return entryPointHandlers
+}
+
+func (m *Manager) buildEntryPointHandler(ctx context.Context, configs map[string]*runtime.UDPRouterInfo) (udp.Handler, error) {
+	logger := log.FromContext(ctx)
+
+	if len(configs) > 1 {
+		logger.Error("Warning: config has more than one udp router for a given entrypoint")
```
Comment (dtomcej) [inline pkg/server/router/udp/router.go]: Why is this warning logged?

Comment (dtomcej) [inline pkg/server/router/udp/router.go]: Why is this warning logged?

File: docs/content/routing/routers/index.md [hunk 2]
```diff
@@ -946,3 +946,160 @@ tcp:
...
+	(that is hardcoded to 3 seconds for now).
+	Making this timeout configurable will be considered later if we get more usage feedback on this matter.
+
+### EntryPoints
+
+If not specified, UDP routers will accept packets from all defined (UDP) entry points.
+If one wants to limit the router scope to a set of entry points, one should set the entry points option.
+
+??? example "Listens to Every Entry Point"
+
```
[the remark this answers is not in the document]
Comment (mpl) [reply docs/content/routing/routers/index.md]: yes. but I don't care that much either way. It is the same example in HTTP and TCP, and we copied it as is for UDP. Maybe there's a reason it was done that way for HTTP and TCP?

File: pkg/server/router/udp/router.go (hunk 1 above)
Comment (mpl) [reply pkg/server/router/udp/router.go]: done

File: cmd/traefik/traefik.go [hunk 3]
```diff
@@ -177,14 +177,19 @@ func setupServer(staticConfiguration *static.Configuration) (*server.Server, err
...
+
 	ctx := context.Background()
 	routinesPool := safe.NewPool(ctx)
 
 	metricsRegistry := registerMetricClients(staticConfiguration.Metrics)
 	accessLog := setupAccessLog(staticConfiguration.AccessLog)
 	chainBuilder := middleware.NewChainBuilder(*staticConfiguration, metricsRegistry, accessLog)
 	managerFactory := service.NewManagerFactory(*staticConfiguration, routinesPool, metricsRegistry)
-	tcpRouterFactory := server.NewTCPRouterFactory(*staticConfiguration, managerFactory, tlsManager, chainBuilder)
+	tcpRouterFactory := server.NewRouterFactory(*staticConfiguration, managerFactory, tlsManager, chainBuilder)
```
[the remark this answers is not in the document]
Comment (mpl) [reply cmd/traefik/traefik.go]: done

File: cmd/traefik/traefik.go [hunk 4]
```diff
@@ -198,7 +203,7 @@ func setupServer(staticConfiguration *static.Configuration) (*server.Server, err
 		metricsRegistry.LastConfigReloadSuccessGauge().Set(float64(time.Now().Unix()))
 	})
 
-	watcher.AddListener(switchRouter(tcpRouterFactory, acmeProviders, serverEntryPointsTCP))
+	watcher.AddListener(switchRouter(tcpRouterFactory, acmeProviders, serverEntryPointsTCP, serverEntryPointsUDP))
```
[the remark this answers is not in the document]
Comment (mpl) [reply cmd/traefik/traefik.go]: done

File: integration/fixtures/udp/wrr.toml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  checkNewVersion = false
+  sendAnonymousUsage = false
+
+[log]
+  level = "DEBUG"
+
+[entryPoints]
+  [entryPoints.udp]
+    address = ":8093/udp"
+  [entryPoints.http]
```
[the remark this answers is not in the document]
Comment (mpl) [reply integration/fixtures/udp/wrr.toml]: done

File: integration/fixtures/udp/wrr.toml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    [udp.services.whoami-b.loadBalancer]
+      [[udp.services.whoami-b.loadBalancer.servers]]
+        address = "{{ .WhoamiBIP}}:8080"
+
+[http]
+  [http.routers]
+    [http.routers.to-whoami-d]
+      service = "whoami"
+      entryPoints = [ "http" ]
```
[the remark this answers is not in the document]
Comment (mpl) [reply integration/fixtures/udp/wrr.toml]: done

**Comment by FL3SH** [comment]
In the current state, can I use it in Kubernetes?

**Comment by jlegido** [comment]
First of all many thanks for your time, and congratulations, since this is a exciting feature.

My question is which traefik version will include this UDP routing feature. Checking [changelog](https://github.com/containous/traefik/compare/v2.1.3...v2.1.4) version [2.1.4](https://github.com/containous/traefik/tree/v2.1.4) looks like it doesn't support it yet.

Maybe will be included in next v2.1.5?

Thanks.

**Comment by ldez** [comment]
You can see the version in the milestone field.

https://github.com/containous/traefik#release-cycle

We respect semver so on v2.1 we only accept bug fixes.

The feature are always in the next minor version.

**Comment by jlegido** [comment]
Many thanks for your quick reply

**Pull request merged**